### DISCUSSION TITLE
USB

### DISCUSSION THREAD

**Discussion by mj-sakellaropoulos** [hansemro/kc1-linux#20] [Q&A] [answered]
Device does not enumerate, as far as I can tell, it's not crashed.

Any special cmdline/modules/kconfigs needed for usb ?

**Comment by hansemro** [comment]
Add this to `/etc/inittab`:
```
ttyGS0::respawn:/sbin/getty -L ttyGS0 115200 vt102
```

<details>
<summary>USB Gadget kconfig</summary>
<a href="https://github.com/hansemro/kc1-linux/blob/75996853dcc142ccd7d2c69e6e28bced4f53e1eb/config/omap4_kc1.config#L3218-L3446">From my kernel .config</a> 
<pre>
CONFIG_USB_GADGET=y
CONFIG_USB_GADGET_DEBUG=y
# CONFIG_USB_GADGET_VERBOSE is not set
CONFIG_USB_GADGET_DEBUG_FILES=y
CONFIG_USB_GADGET_DEBUG_FS=y
CONFIG_USB_GADGET_VBUS_DRAW=500
CONFIG_USB_GADGET_STORAGE_NUM_BUFFERS=2
# CONFIG_U_SERIAL_CONSOLE is not set
#
# USB Gadget precomposed configurations
#
# CONFIG_USB_ZERO is not set
# CONFIG_USB_ETH is not set
# CONFIG_USB_G_NCM is not set
# CONFIG_USB_GADGETFS is not set
# CONFIG_USB_FUNCTIONFS is not set
# CONFIG_USB_MASS_STORAGE is not set
# CONFIG_USB_G_SERIAL is not set
# CONFIG_USB_G_PRINTER is not set
CONFIG_USB_CDC_COMPOSITE=y
# CONFIG_USB_G_NOKIA is not set
# CONFIG_USB_G_ACM_MS is not set
# CONFIG_USB_G_MULTI is not set
# CONFIG_USB_G_HID is not set
# CONFIG_USB_G_DBGP is not set
# CONFIG_USB_RAW_GADGET is not set
# end of USB Gadget precomposed configurations
</pre>
</details>

<details>
<summary>dmesg</summary>
<pre>
[    4.076110] usbcore: registered new interface driver cdc_acm
[    4.082885] cdc_acm: USB Abstract Control Model driver for USB modems and ISDN adapters
[    4.092285] usbcore: registered new interface driver cdc_wdm
[    4.193572] udc musb-hdrc.0.auto: registering UDC driver [g_cdc]
[    4.213470] g_cdc gadget: adding config #1 'CDC Composite (ECM + ACM)'/(ptrval)
[    4.222137] g_cdc gadget: adding 'cdc_ethernet'/(ptrval) to config 'CDC Composite (ECM + ACM)'/(ptrval)
[    4.245391] g_cdc gadget: CDC Ethernet: dual speed IN/ep1in OUT/ep1out NOTIFY/ep2in
[    4.254272] g_cdc gadget: adding 'acm'/(ptrval) to config 'CDC Composite (ECM + ACM)'/(ptrval)
[    4.264129] g_cdc gadget: acm ttyGS0: dual speed IN/ep3in OUT/ep2out NOTIFY/ep4in
[    4.272827] g_cdc gadget: cfg 1/(ptrval) speeds: high full
[    4.279388] g_cdc gadget:   interface 0 = cdc_ethernet/(ptrval)
[    4.286376] g_cdc gadget:   interface 1 = cdc_ethernet/(ptrval)
[    4.293395] g_cdc gadget:   interface 2 = acm/(ptrval)
[    4.299621] g_cdc gadget:   interface 3 = acm/(ptrval)
[    4.305786] g_cdc gadget: CDC Composite Gadget, version: King Kamehameha Day 2008
[    4.314453] g_cdc gadget: g_cdc ready
[    7.745697] g_cdc gadget: high-speed config #1: CDC Composite (ECM + ACM)
[    7.749237] g_cdc gadget: init ecm
[    7.758575] g_cdc gadget: notify connect false
[    7.764068] g_cdc gadget: reset acm ttyGS0
[    7.769134] g_cdc gadget: activate acm ttyGS0
[    7.774444] g_cdc gadget: acm ttyGS0 serial state 0000
[    7.781494] g_cdc gadget: notify speed 425984000
[    7.787445] g_cdc gadget: activate ecm
[    7.796112] g_cdc gadget: ecm_close
[    7.801208] g_cdc gadget: packet filter 0c
[    7.801208] g_cdc gadget: ecm req21.43 v000c i0000 l0
[    7.814270] g_cdc gadget: acm ttyGS0 req21.20 v0000 i0002 l7
[    8.394836] g_cdc gadget: packet filter 0e
[    8.399932] g_cdc gadget: ecm req21.43 v000e i0000 l0
[    8.445190] g_cdc gadget: notify connect false
[    8.450927] g_cdc gadget: notify speed 425984000
[    9.880889] g_cdc gadget: acm ttyGS0 req21.22 v0003 i0002 l0
[    9.887664] g_cdc gadget: acm ttyGS0 req21.20 v0000 i0002 l7
[    9.894592] g_cdc gadget: acm ttyGS0 req21.22 v0002 i0002 l0
[   10.001464] g_cdc gadget: acm ttyGS0 req21.22 v0003 i0002 l0
[   14.425903] g_cdc gadget: packet filter 0e
[   14.425903] g_cdc gadget: ecm req21.43 v000e i0000 l0
[   20.965911] g_cdc gadget: acm ttyGS0 req21.22 v0000 i0002 l0
[   21.073059] g_cdc gadget: acm ttyGS0 req21.22 v0003 i0002 l0
[   21.073303] g_cdc gadget: acm ttyGS0 req21.22 v0002 i0002 l0
[   21.173706] g_cdc gadget: acm ttyGS0 req21.22 v0003 i0002 l0
[   30.331604] g_cdc gadget: ecm_open
[   30.331604] g_cdc gadget: notify connect true
[   30.331604] g_cdc gadget: notify speed 425984000
[   30.334594] g_cdc gadget: packet filter 0e
[   30.334594] g_cdc gadget: ecm req21.43 v000e i0000 l0
[   30.629455] g_cdc gadget: acm ttyGS0 serial state 0003
[   30.654235] g_cdc gadget: acm ttyGS0 serial state 0003
[   30.960968] g_cdc gadget: acm ttyGS0 req21.22 v0000 i0002 l0
[   31.014434] g_cdc gadget: acm ttyGS0 req21.22 v0003 i0002 l0
[   31.014953] g_cdc gadget: acm ttyGS0 req21.20 v0000 i0002 l7
[   36.963439] g_cdc gadget: acm ttyGS0 req21.22 v0000 i0002 l0
</pre>
</details>

On my PC, I see /dev/ACM0 get enumerated with the device showing up in lsusb as:
```
Bus 001 Device 094: ID 0525:a4aa Netchip Technology, Inc. Linux-USB CDC Composite Gadge (Ethernet and ACM)
```

To connect (with picocom, for example): `picocom /dev/ttyACM0 -b 115200`

**Reply by hansemro** [reply]
```
$ picocom /dev/ttyACM0 -b 115200
picocom v3.1

port is        : /dev/ttyACM0
flowcontrol    : none
baudrate is    : 115200
parity is      : none
databits are   : 8
stopbits are   : 1
escape is      : C-a
local echo is  : no
noinit is      : no
noreset is     : no
hangup is      : no
nolock is      : no
send_cmd is    : sz -vv
receive_cmd is : rz -vv -E
imap is        : 
omap is        : 
emap is        : crcrlf,delbs,
logfile is     : none
initstring     : none
exit_after is  : not set
exit is        : no

Type [C-a] [C-h] to see available commands
Terminal ready

Password: 
Welcome to postmarketOS
Kernel 5.11.0+ on an armv7l (/dev/ttyGS0)
amazon-otter login: root
Password: 
Welcome to postmarketOS! o/

This distribution is based on Alpine Linux.
First time using postmarketOS? Make sure to read the cheatsheet in the wiki:

-> https://postmarketos.org/cheatsheet

You may change this message by editing /etc/motd.
amazon-otter:~# uname -a
Linux amazon-otter 5.11.0+ #35 SMP Thu Apr 22 21:49:52 PDT 2021 armv7l GNU/Linux
amazon-otter:~# 

```

**Comment by mj-sakellaropoulos** [comment]
Weird, I already have those lines in defconfig. I can see the gadget drivers are loaded, but when I unplug/replug nothing happens on either end. I have this warning in startup log:
```
[    5.572692] udc-core: couldn't find an available UDC - added [g_cdc] to list of pending drivers
```

**Reply by hansemro** [reply]
You may be missing musb/otg/dualrole (?) because I see this:
```
[    3.931335] udc musb-hdrc.0.auto: registering UDC driver [g_cdc]
```

**Reply by mj-sakellaropoulos** [reply]
AFAIK, the UDC error means there is no available host port. Seems to be device tree related, but my device tree usb_otg entry is the same as yours. Weird.

**Reply by hansemro** [reply]
Can you send me your current kernel config + dtb. I want to try it on my end.

**Reply by hansemro** [reply]
My previously-built image
[2021-04-23-07-40_5.11.0.tar.gz](https://github.com/hansemro/kc1-linux/files/6366337/2021-04-23-07-40_5.11.0.tar.gz)
[dts.tar.gz](https://github.com/hansemro/kc1-linux/files/6366418/dts.tar.gz)

**Reply by mj-sakellaropoulos** [reply]
kernel config is as-is on on my repo, I will update the dtb soon
https://github.com/mj-sakellaropoulos/kc1_linux_configs/tree/main/drm

**Reply by mj-sakellaropoulos** [reply]
> My previously-built image
> [2021-04-23-07-40_5.11.0.tar.gz](https://github.com/hansemro/kc1-linux/files/6366337/2021-04-23-07-40_5.11.0.tar.gz)
> [dts.tar.gz](https://github.com/hansemro/kc1-linux/files/6366418/dts.tar.gz)

Ok, I can confirm your kernel works 1000000x better than mine ever did lol. I managed to boot my old debian partition. gadget is functional.

**Comment by mj-sakellaropoulos** [comment]
Apparently this is _not_ a devicetree issue, because I replaced the devicetree in your boot.img with mine and USB is still working. Looks like a kernel issue now.

**Reply by hansemro** [reply]
Yeah, looks like a kernel config issue. I recompiled 5.10 source (using buildroot) with my 5.11.0 config and it reaches userspace.

[bad log.txt](https://github.com/hansemro/kc1-linux/files/6366885/log.txt)

**Comment by mj-sakellaropoulos** [comment, marked as answer]
Finally... via cdc_acm
Used the config in the tar archive above
```
# dmesg | grep panel
[    5.654785] panel-simple display: GPIO lookup for consumer hpd
[    5.661041] panel-simple display: using device tree for GPIO lookup
[    5.685546] panel-simple display: using lookup tables for GPIO lookup
[    5.692413] panel-simple display: No GPIO consumer hpd found
[    5.698425] panel-simple display: Looking up power-supply from device tree
[    8.340942] panel-simple display: GPIO lookup for consumer hpd
[    8.340942] panel-simple display: using device tree for GPIO lookup
[    8.371734] panel-simple display: using lookup tables for GPIO lookup
[    8.378601] panel-simple display: No GPIO consumer hpd found
[    8.384582] panel-simple display: Looking up power-supply from device tree
[    8.391876] panel-simple display: GPIO lookup for consumer enable
[    8.398956] panel-simple display: using device tree for GPIO lookup
[    8.422271] panel-simple display: Specify missing connector_type
# dmesg | grep dss
[    5.616485] omapdss_dss 58000000.dss: Looking up vdda_video-supply from device tree
[    5.624786] omapdss_dss 58000000.dss: Looking up vdda_video-supply property in node /ocp/target-module@58000000/dss@0 failed
[    5.636688] omapdss_dss 58000000.dss: supply vdda_video not found, using dummy regulator
[    8.308197] omapdss_dss 58000000.dss: Looking up vdda_video-supply from device tree
[    8.308197] omapdss_dss 58000000.dss: Looking up vdda_video-supply property in node /ocp/target-module@58000000/dss@0 failed
[    8.328430] omapdss_dss 58000000.dss: supply vdda_video not found, using dummy regulator
[    8.434631] pinctrl-single 4a100040.pinmux: pin PIN145 already requested by display; cannot claim for 58000000.dss
[    8.446258] pinctrl-single 4a100040.pinmux: pin-145 (58000000.dss) status -22
[    8.460205] pinctrl-single 4a100040.pinmux: could not request pin 145 (PIN145) from group pinmux_dss_dpi_pins  on device pinctrl-single
[    8.480834] omapdss_dss 58000000.dss: Error applying setting, reverse things back
[    8.507263] omapdss_dss: probe of 58000000.dss failed with error -22
# uname -a
Linux buildroot 5.10.25 #4 SMP Fri Apr 23 15:34:09 EDT 2021 armv7l GNU/Linux
```

**Comment by hansemro** [comment]
I am not sure why, but usb CDC/ACM is pretty flaky on 5.12.X. @mj-sakellaropoulos Do you have this issue as well?

**Reply by mj-sakellaropoulos** [reply]
Yup. Bigtime.

For me it's the ethernet gadget that is the least stable.

**Reply by hansemro** [reply]
https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=v5.12-rc6&id=de879a8d37ff86b8111f94a64fc82c81c1dddd89 seems to introduce the regression.

**Reply by hansemro** [reply]
https://patchwork.kernel.org/project/linux-omap/patch/20210324071142.42264-1-tony@atomide.com/ specifically introduces the regression

**Reply by hansemro** [reply]
[This patch](https://lore.kernel.org/linux-usb/YLENtd45ly8ZFJO2@piout.net/T/) seems to fix the USB gadget reset issue.
